File: gcp-api/lib/print_processor.py
"""
Print processor module for passport photo printing.
Handles arranging processed photos for printing on standard 10×15 cm paper.
"""

import cv2
import numpy as np
from lib.config import Config

class PrintProcessor:
    """Handles arranging passport photos for printing with cutting guides."""
    
    def __init__(self):
        """Initialize the PrintProcessor."""
        self.config = Config()
        
        # Print paper dimensions (10×15 cm)
        self.PAPER_WIDTH_MM = 100
        self.PAPER_HEIGHT_MM = 150
        
        # Calculate paper dimensions in pixels at target DPI
        self.PAPER_WIDTH_PX = int((self.PAPER_WIDTH_MM / 25.4) * self.config.TARGET_DPI)
        self.PAPER_HEIGHT_PX = int((self.PAPER_HEIGHT_MM / 25.4) * self.config.TARGET_DPI)
        
        # Photo dimensions (already defined in config)
        self.PHOTO_WIDTH_PX = self.config.FINAL_OUTPUT_WIDTH_PX
        self.PHOTO_HEIGHT_PX = self.config.FINAL_OUTPUT_HEIGHT_PX
        
        # Grid layout (2x2 for 4 photos)
        self.GRID_COLS = 2
        self.GRID_ROWS = 2
        
        # Calculate spacing and margins
        self._calculate_layout()
    
    def _calculate_layout(self):
        """Calculate the optimal layout for photos on the print paper."""
        # Calculate total area needed for photos
        total_photos_width = self.PHOTO_WIDTH_PX * self.GRID_COLS
        total_photos_height = self.PHOTO_HEIGHT_PX * self.GRID_ROWS
        
        # Calculate margins
        self.margin_x = (self.PAPER_WIDTH_PX - total_photos_width) // 2
        self.margin_y = (self.PAPER_HEIGHT_PX - total_photos_height) // 2
        
        # Ensure minimum margins
        min_margin_px = int((5 / 25.4) * self.config.TARGET_DPI)  # 5mm minimum margin
        self.margin_x = max(self.margin_x, min_margin_px)
        self.margin_y = max(self.margin_y, min_margin_px)
        
        # Calculate spacing between photos
        self.spacing_x = 0  # No spacing between photos for easier cutting
        self.spacing_y = 0
    
    def _draw_cutting_guides(self, print_canvas):
        """Draw cutting guides on the print canvas."""
        # Cutting guide color (light gray)
        guide_color = (128, 128, 128)
        guide_thickness = 2
        
        # Draw outer cutting frame
        frame_margin = int((2 / 25.4) * self.config.TARGET_DPI)  # 2mm outside the photos
        
        # Calculate frame coordinates
        frame_x1 = self.margin_x - frame_margin
        frame_y1 = self.margin_y - frame_margin
        frame_x2 = self.margin_x + (self.PHOTO_WIDTH_PX * self.GRID_COLS) + frame_margin
        frame_y2 = self.margin_y + (self.PHOTO_HEIGHT_PX * self.GRID_ROWS) + frame_margin
        
        # Ensure frame is within canvas bounds
        frame_x1 = max(0, frame_x1)
        frame_y1 = max(0, frame_y1)
        frame_x2 = min(self.PAPER_WIDTH_PX, frame_x2)
        frame_y2 = min(self.PAPER_HEIGHT_PX, frame_y2)
        
        # Draw outer frame
        cv2.rectangle(print_canvas, (frame_x1, frame_y1), (frame_x2, frame_y2), guide_color, guide_thickness)
        
        # Draw individual photo cutting guides
        for row in range(self.GRID_ROWS):
            for col in range(self.GRID_COLS):
                x = self.margin_x + col * self.PHOTO_WIDTH_PX
                y = self.margin_y + row * self.PHOTO_HEIGHT_PX
                
                # Draw photo border
                cv2.rectangle(print_canvas, 
                            (x, y), 
                            (x + self.PHOTO_WIDTH_PX, y + self.PHOTO_HEIGHT_PX), 
                            guide_color, 1)
        
        # Draw cross marks at corners for precise cutting
        cross_size = int((3 / 25.4) * self.config.TARGET_DPI)  # 3mm cross marks
        
        # Corner positions for each photo
        for row in range(self.GRID_ROWS + 1):  # +1 to include bottom edge
            for col in range(self.GRID_COLS + 1):  # +1 to include right edge
                x = self.margin_x + col * self.PHOTO_WIDTH_PX
                y = self.margin_y + row * self.PHOTO_HEIGHT_PX
                
                # Draw cross mark
                cv2.line(print_canvas, 
                        (x - cross_size, y), (x + cross_size, y), 
                        guide_color, 1)
                cv2.line(print_canvas, 
                        (x, y - cross_size), (x, y + cross_size), 
                        guide_color, 1)
    
    def _add_print_info(self, print_canvas):
        """Add printing information text to the canvas."""
        # Font settings
        font = cv2.FONT_HERSHEY_SIMPLEX
        font_scale = 0.6
        font_color = (64, 64, 64)  # Dark gray
        thickness = 1
        
        # Calculate text position (bottom of canvas)
        text_y = self.PAPER_HEIGHT_PX - int((3 / 25.4) * self.config.TARGET_DPI)  # 3mm from bottom
        
        # Add dimension info
        info_text = f"4x Passport Photos (35x45mm) - Cut along guides"
        text_size = cv2.getTextSize(info_text, font, font_scale, thickness)[0]
        text_x = (self.PAPER_WIDTH_PX - text_size[0]) // 2
        
        cv2.putText(print_canvas, info_text, (text_x, text_y), 
                   font, font_scale, font_color, thickness)
        
        # Add DPI info
        dpi_text = f"Print at {self.config.TARGET_DPI} DPI"
        dpi_size = cv2.getTextSize(dpi_text, font, font_scale, thickness)[0]
        dpi_x = (self.PAPER_WIDTH_PX - dpi_size[0]) // 2
        dpi_y = text_y + int((4 / 25.4) * self.config.TARGET_DPI)  # 4mm below main text
        
        cv2.putText(print_canvas, dpi_text, (dpi_x, dpi_y), 
                   font, font_scale, font_color, thickness)
    
    def create_print_layout(self, photo_bgr):
        """
        Create a print layout with 4 copies of the photo arranged on 10×15 cm paper.
        
        Args:
            photo_bgr (numpy.ndarray): The processed passport photo in BGR format
            
        Returns:
            tuple: (print_canvas, info_dict) where print_canvas is the printable image
                   and info_dict contains layout information
        """
        if photo_bgr is None:
            raise ValueError("Photo cannot be None")
        
        # Validate photo dimensions
        if photo_bgr.shape[:2] != (self.PHOTO_HEIGHT_PX, self.PHOTO_WIDTH_PX):
            # Resize photo to expected dimensions if needed
            photo_bgr = cv2.resize(photo_bgr, (self.PHOTO_WIDTH_PX, self.PHOTO_HEIGHT_PX), 
                                 interpolation=cv2.INTER_AREA)
        
        # Create white canvas for print layout
        print_canvas = np.full((self.PAPER_HEIGHT_PX, self.PAPER_WIDTH_PX, 3), 
                              255, dtype=np.uint8)
        
        # Place 4 copies of the photo in a 2×2 grid
        for row in range(self.GRID_ROWS):
            for col in range(self.GRID_COLS):
                # Calculate position for this photo
                x = self.margin_x + col * (self.PHOTO_WIDTH_PX + self.spacing_x)
                y = self.margin_y + row * (self.PHOTO_HEIGHT_PX + self.spacing_y)
                
                # Place photo on canvas
                print_canvas[y:y+self.PHOTO_HEIGHT_PX, x:x+self.PHOTO_WIDTH_PX] = photo_bgr
        
        # Add cutting guides
        self._draw_cutting_guides(print_canvas)
        
        # Add print information
        self._add_print_info(print_canvas)
        
        # Create info dictionary
        info_dict = {
            "paper_size_mm": f"{self.PAPER_WIDTH_MM}×{self.PAPER_HEIGHT_MM}",
            "paper_size_px": f"{self.PAPER_WIDTH_PX}×{self.PAPER_HEIGHT_PX}",
            "photo_size_mm": f"{self.config.TARGET_PHOTO_WIDTH_MM}×{self.config.TARGET_PHOTO_HEIGHT_MM}",
            "photo_size_px": f"{self.PHOTO_WIDTH_PX}×{self.PHOTO_HEIGHT_PX}",
            "photos_count": self.GRID_ROWS * self.GRID_COLS,
            "dpi": self.config.TARGET_DPI,
            "margins_px": {"x": self.margin_x, "y": self.margin_y}
        }
        
        return print_canvas, info_dict
    
    def save_print_layout(self, photo_bgr, output_path):
        """
        Create and save a print layout to a file.
        
        Args:
            photo_bgr (numpy.ndarray): The processed passport photo in BGR format
            output_path (str): Path to save the print layout
            
        Returns:
            dict: Layout information
        """
        print_canvas, info_dict = self.create_print_layout(photo_bgr)
        
        # Save with high quality
        cv2.imwrite(output_path, print_canvas, [cv2.IMWRITE_JPEG_QUALITY, 95])
        
        return info_dict 

File: server/print-processor.ts
/**
 * Print processor module for passport photo printing.
 * Arranges processed photos for printing on standard 10×15 cm paper.
 */

import sharp from 'sharp'
import { ICAOConfig, COUNTRY_DIMENSIONS } from './validation-constants.js'

// Paper and layout constants
const PAPER_WIDTH_MM = 100
const PAPER_HEIGHT_MM = 150
const PHOTO_SPACING_MM = 6
const MIN_MARGIN_MM = 5
const BOTTOM_MARGIN_MM = 4
const LINE_SPACING_MM = 2
const TICK_LENGTH_MM = 6

// Grid layout
const GRID_COLS = 2
const GRID_ROWS = 2

// Convert mm to pixels at target DPI
function mmToPx(mm: number): number {
    return Math.floor((mm / 25.4) * ICAOConfig.targetDpi)
}

// Layout dimensions in pixels
const PAPER_WIDTH_PX = mmToPx(PAPER_WIDTH_MM)
const PAPER_HEIGHT_PX = mmToPx(PAPER_HEIGHT_MM)
const SPACING_PX = Math.max(mmToPx(PHOTO_SPACING_MM), 1)
const MIN_MARGIN_PX = mmToPx(MIN_MARGIN_MM)
const TICK_LENGTH_PX = Math.max(mmToPx(TICK_LENGTH_MM), 6)

// Calculate layout margins
function calculateLayout(photoWidthPx: number, photoHeightPx: number) {
    const totalPhotosWidth = photoWidthPx * GRID_COLS + SPACING_PX * (GRID_COLS - 1)
    const totalPhotosHeight = photoHeightPx * GRID_ROWS + SPACING_PX * (GRID_ROWS - 1)

    const marginX = Math.max(Math.floor((PAPER_WIDTH_PX - totalPhotosWidth) / 2), MIN_MARGIN_PX)
    const marginY = Math.max(Math.floor((PAPER_HEIGHT_PX - totalPhotosHeight) / 2), MIN_MARGIN_PX)

    return { marginX, marginY }
}

/**
 * Creates an SVG overlay with cutting guides and print info.
 */
function createOverlaySvg(marginX: number, marginY: number, photoWidthPx: number, photoHeightPx: number, countryCode: string, docType: string): string {
    const guideColor = '#808080'
    const guideThickness = 2
    const fontColor = '#404040'

    const lines: string[] = []

    // Draw cutting guides for each photo position
    for (let row = 0; row < GRID_ROWS; row++) {
        for (let col = 0; col < GRID_COLS; col++) {
            const x0 = marginX + col * (photoWidthPx + SPACING_PX)
            const y0 = marginY + row * (photoHeightPx + SPACING_PX)
            const x1 = x0 + photoWidthPx
            const y1 = y0 + photoHeightPx

            // Top-left corner
            lines.push(`<line x1="${Math.max(x0 - TICK_LENGTH_PX, 0)}" y1="${y0}" x2="${x0}" y2="${y0}" stroke="${guideColor}" stroke-width="${guideThickness}"/>`)
            lines.push(`<line x1="${x0}" y1="${Math.max(y0 - TICK_LENGTH_PX, 0)}" x2="${x0}" y2="${y0}" stroke="${guideColor}" stroke-width="${guideThickness}"/>`)

            // Top-right corner
            lines.push(`<line x1="${x1}" y1="${y0}" x2="${Math.min(x1 + TICK_LENGTH_PX, PAPER_WIDTH_PX)}" y2="${y0}" stroke="${guideColor}" stroke-width="${guideThickness}"/>`)
            lines.push(`<line x1="${x1}" y1="${Math.max(y0 - TICK_LENGTH_PX, 0)}" x2="${x1}" y2="${y0}" stroke="${guideColor}" stroke-width="${guideThickness}"/>`)

            // Bottom-left corner
            lines.push(`<line x1="${Math.max(x0 - TICK_LENGTH_PX, 0)}" y1="${y1}" x2="${x0}" y2="${y1}" stroke="${guideColor}" stroke-width="${guideThickness}"/>`)
            lines.push(`<line x1="${x0}" y1="${y1}" x2="${x0}" y2="${Math.min(y1 + TICK_LENGTH_PX, PAPER_HEIGHT_PX)}" stroke="${guideColor}" stroke-width="${guideThickness}"/>`)

            // Bottom-right corner
            lines.push(`<line x1="${x1}" y1="${y1}" x2="${Math.min(x1 + TICK_LENGTH_PX, PAPER_WIDTH_PX)}" y2="${y1}" stroke="${guideColor}" stroke-width="${guideThickness}"/>`)
            lines.push(`<line x1="${x1}" y1="${y1}" x2="${x1}" y2="${Math.min(y1 + TICK_LENGTH_PX, PAPER_HEIGHT_PX)}" stroke="${guideColor}" stroke-width="${guideThickness}"/>`)
        }
    }

    // Add print info text at bottom
    const bottomMarginPx = mmToPx(BOTTOM_MARGIN_MM)
    const lineSpacingPx = mmToPx(LINE_SPACING_MM)
    const fontSize = Math.floor(ICAOConfig.targetDpi / 15) // Scale font with DPI

    let typeLabel = docType === 'drivers_license' ? 'Driving License' : 'Passport';
    let dimsStr = '35x45mm';

    if (countryCode && COUNTRY_DIMENSIONS[countryCode.toUpperCase()]) {
        const docKey = docType === 'drivers_license' ? 'drivers_license' : 'passport';
        const dims = COUNTRY_DIMENSIONS[countryCode.toUpperCase()][docKey];
        if (dims) {
            dimsStr = `${dims.widthMm}x${dims.heightMm}mm`;
        }
    }

    const urlText = 'www.passportphotovalidator.com'
    const infoText = `4x ${countryCode.toUpperCase()} ${typeLabel} Photos (${dimsStr}) - Cut along guides`

    const infoY = PAPER_HEIGHT_PX - bottomMarginPx
    const urlY = infoY - fontSize - lineSpacingPx

    lines.push(`<text x="${PAPER_WIDTH_PX / 2}" y="${urlY}" text-anchor="middle" font-family="sans-serif" font-size="${fontSize}" font-weight="bold" fill="${fontColor}">${urlText}</text>`)
    lines.push(`<text x="${PAPER_WIDTH_PX / 2}" y="${infoY}" text-anchor="middle" font-family="sans-serif" font-size="${Math.floor(fontSize * 0.8)}" fill="${fontColor}">${infoText}</text>`)

    return `<svg width="${PAPER_WIDTH_PX}" height="${PAPER_HEIGHT_PX}" xmlns="http://www.w3.org/2000/svg">
    ${lines.join('\n    ')}
  </svg>`
}

export interface PrintLayoutResult {
    success: boolean
    printImage?: Buffer
    metadata?: {
        paperSizeMm: string
        photosCount: number
        paperWidthPx: number
        paperHeightPx: number
    }
    error?: string
}

/**
 * Creates a print layout with 4 passport photos arranged in a 2x2 grid
 * on 10x15cm paper with cutting guides.
 *
 * @param photoBuffer - The processed passport photo as a PNG or WebP buffer
 * @param countryCode - The target country code (e.g. 'BE', 'US')
 * @param docType - The document type ('passport' or 'drivers_license')
 * @returns PrintLayoutResult with the print-ready image
 */
export async function createPrintLayout(
    photoBuffer: Buffer,
    countryCode: string = 'BE',
    docType: string = 'passport'
): Promise<PrintLayoutResult> {
    try {

        let widthMm: number = ICAOConfig.targetPhotoWidthMm;
        let heightMm: number = ICAOConfig.targetPhotoHeightMm;

        if (countryCode && COUNTRY_DIMENSIONS[countryCode.toUpperCase()]) {
            const docKey = docType === 'drivers_license' ? 'drivers_license' : 'passport';
            const dims = COUNTRY_DIMENSIONS[countryCode.toUpperCase()][docKey];
            if (dims) {
                widthMm = dims.widthMm;
                heightMm = dims.heightMm;
            }
        }

        const photoWidthPx = mmToPx(widthMm);
        const photoHeightPx = mmToPx(heightMm);

        const { marginX, marginY } = calculateLayout(photoWidthPx, photoHeightPx)

        // Ensure the photo is the correct size
        // Flatten to white background and use PNG for strictly lossless quality
        const resizedPhoto = await sharp(photoBuffer)
            .resize(photoWidthPx, photoHeightPx, { fit: 'fill' })
            .flatten({ background: { r: 255, g: 255, b: 255 } })
            .png()
            .toBuffer()

        // Build composite operations for placing photos in a 2x2 grid
        const compositeOperations: sharp.OverlayOptions[] = []

        for (let row = 0; row < GRID_ROWS; row++) {
            for (let col = 0; col < GRID_COLS; col++) {
                const x = marginX + col * (photoWidthPx + SPACING_PX)
                const y = marginY + row * (photoHeightPx + SPACING_PX)

                compositeOperations.push({
                    input: resizedPhoto,
                    left: x,
                    top: y,
                })
            }
        }

        // Create the SVG overlay with cutting guides
        const overlaySvg = createOverlaySvg(marginX, marginY, photoWidthPx, photoHeightPx, countryCode, docType)
        const overlayBuffer = Buffer.from(overlaySvg)

        compositeOperations.push({
            input: overlayBuffer,
            left: 0,
            top: 0,
        })

        // Create white canvas and composite all elements
        const printImage = await sharp({
            create: {
                width: PAPER_WIDTH_PX,
                height: PAPER_HEIGHT_PX,
                channels: 3,
                background: { r: 255, g: 255, b: 255 },
            },
        })
            .composite(compositeOperations)
            .withMetadata({ density: 600 })
            .png()
            .toBuffer()

        return {
            success: true,
            printImage,
            metadata: {
                paperSizeMm: `${PAPER_WIDTH_MM}x${PAPER_HEIGHT_MM}`,
                photosCount: GRID_ROWS * GRID_COLS,
                paperWidthPx: PAPER_WIDTH_PX,
                paperHeightPx: PAPER_HEIGHT_PX,
            },
        }
    } catch (error) {
        console.error('Print layout creation error:', error)
        return {
            success: false,
            error: error instanceof Error ? error.message : 'Failed to create print layout',
        }
    }
}
